operators/src/processing/vector_join/equi_data_join.rs

```diff
@@ -0,0 +1,401 @@
+            right_processor,
+            left_column,
+            right_column,
+            right_column_prefix: right_prefix,
+        }
+    }
+
+    fn join<'c>(
+        &self,
+        left: &'c FeatureCollection<G>,
+        right: &DataCollection,
+    ) -> Result<FeatureCollection<G>> {
+        let right = self.fix_columns(right, &left.column_names().collect())?;
+
+        let left_join_column = left.data(&self.left_column)?;
+        let right_join_column = right.data(&self.right_column)?;
+
+        // TODO: use macro
+        match (left_join_column, right_join_column) {
+            (FeatureDataRef::Number(c1), FeatureDataRef::Number(c2)) => {
+                Self::compute_join(&c1, c2, left, &right)
+            }
+            (FeatureDataRef::Decimal(c1), FeatureDataRef::Decimal(c2)) => {
+                Self::compute_join(&c1, c2, left, &right)
+            }
+            (FeatureDataRef::Text(c1), FeatureDataRef::Text(c2)) => {
+                Self::compute_join(&c1, c2, left, &right)
+            }
+            (FeatureDataRef::Categorical(c1), FeatureDataRef::Categorical(c2)) => {
+                Self::compute_join(&c1, c2, left, &right)
+            }
+            (left, right) => Err(Error::ColumnTypeMismatch {
+                left: (&left).into(),
+                right: (&right).into(),
+            }),
+        }
+    }
+
+    fn compute_join<'i, D1, D2, T>(
+        left_join_values: &D1,
+        right_join_values: D2,
+        left: &FeatureCollection<G>,
+        right: &DataCollection,
+    ) -> Result<FeatureCollection<G>>
+    where
+        D1: DataRef<'i, T> + 'i,
+        D2: DataRef<'i, T> + 'i,
+        T: 'i + PartialEq,
+    {
+        let matches_iter = left_join_values
+            .as_ref()
+            .iter()
+            .zip(left.time_intervals())
+            .map(move |(left_value, left_time_interval)| {
+                right_join_values
+                    .as_ref()
+                    .iter()
+                    .zip(right.time_intervals())
+                    .map(|(right_value, right_time_interval)| {
+                        (left_value == right_value)
+                            && left_time_interval.intersects(right_time_interval)
+                    })
+                    .collect::<Vec<bool>>()
+            });
+
+        let mut builder = FeatureCollection::<G>::builder();
+
+        // create header by combining values from both collections
+        for (column_name, column_type) in left
+            .column_types()
+            .into_iter()
+            .chain(right.column_types().into_iter())
+        {
+            builder.add_column(column_name, column_type)?;
+        }
+
+        let mut builder = builder.finish_header();
+
+        for ((left_feature_idx, matches), geometry) in matches_iter
+            .enumerate()
+            .zip(left.geometries().map(Into::into))
+        {
+            // TODO: is it better to traverse column-wise?
```

Comment:
I think so too because you are performing row-based operations on a columnar data layout.

Reply:
building the feature collection this way is quite row oriented, but we can try to batch-insert column-wise

dbfdf79